crucible-symio/src/Lang/Crucible/SymIO/Loader.hs

```diff
@@ -0,0 +1,167 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+-- | This module defines a default loader for initial symbolic filesystem contents
+--
+-- It uses a simple convention to convert on-disk files and metadata into a
+-- 'SymIO.InitialFileSystemContents'. This is not the only way to construct
+-- initial filesystem contents, but it is a good default if a tool does not have
+-- more specific needs.
+--
+-- The caller provides a single input: a path to a directory.  The directory
+-- contains two things:
+--
+-- 1. A subdirectory named @root@ that contains the concrete files in the symbolic filesystem (i.e., the directory mapped to @/@)
+-- 2. (Optional) A file named @symbolic-manifest.json@, which describes symbolic files and overlays
+--
+-- The symbolic manifest specifies the contents of symbolic files, including
+-- constraints on symbolic values.  Furthermore, it enables users to specify
+-- that concrete files in the provided filesystem have symbolic values overlaid
+-- over the concrete values. If an overlay is specified in the symbolic
+-- manifest, the referenced concrete file /must/ exist.
+--
+-- Note: future versions of this interface could support symbolic filesystems
+-- stored in zip or tar files.
+module Lang.Crucible.SymIO.Loader (
+    loadInitialFiles
+  , FileSystemLoadError(..)
+  ) where
+
+import qualified Control.Exception as X
+import qualified Data.Aeson as JSON
+import qualified Data.ByteString as BS
+import qualified Data.Foldable as F
+import qualified Data.List as List
+import qualified Data.Map.Strict as Map
+import           Data.Maybe ( fromMaybe )
+import qualified Data.Parameterized.NatRepr as PN
+import qualified Data.Traversable as T
+import           Data.Word ( Word64 )
+import           GHC.Generics ( Generic )
+import qualified System.Directory as SD
+import           System.FilePath ( (</>) )
+import qualified System.FilePath.Find as SFF
+import qualified What4.BaseTypes as WT
+import qualified What4.Interface as WI
+
+import qualified Lang.Crucible.Backend as LCB
+import qualified Lang.Crucible.SymIO as SymIO
+
+data FileSystemLoadError = ErrorDecodingJSON String
+                         | forall k . FileSpecifiedAsSymbolicAndConcrete (SymIO.FDTarget k)
+
+deriving instance Show FileSystemLoadError
+
+instance X.Exception FileSystemLoadError
+
+-- | The specification for the symbolic contents of a file in the symbolic
+-- filesystem
+--
+-- There will be multiple specifications including:
+--
+--   * Complete symbolic file specifications (including concrete regions)
+--   * Symbolic overlays on otherwise concrete files
+data SymbolicFileContents =
```

Comment:
My inclination is to leave this as data because I'm planning on adding more fields in the near term (e.g., specifying constraints over ranges of symbolic values)